sdk/servicebus/service-bus/review/service-bus.api.md

```diff
@@ -150,8 +150,9 @@ export class Sender {
 
 // @public
 export class ServiceBusClient {
+    constructor(connectionString: string, options?: ServiceBusClientOptions);
+    constructor(credential: TokenCredential, host: string, options?: ServiceBusClientOptions);
```

Comment:
I think the ordering of parameters here should be switched for consistency with event hubs.
```typescript
constructor(host: string, credential: TokenCredential, options?: ServiceBusClientOptions);
```
This also matches the sample in the typescript design [guidelines](https://azure.github.io/azure-sdk/typescript_design.html#ts-apisurface-serviceclient).

Reply:
Oh, could you point to which guideline is being considered for this?

Because I found the following that seems similar / applicable for `hostOrConnectionString` usage

```
// bad example
class ExampleClient {
  constructor (connectionString: string, options: ExampleClientOptions);
  constructor (url: string, options: ExampleClientOptions);
  constructor (urlOrCS: string, options: ExampleClientOptions) {
    // have to dig into the first parameter to see whether its
    // a url or a connection string. Not ideal.
  }
}
```

Reply:
Link was hidden in the word 'guidelines' :)
https://azure.github.io/azure-sdk/typescript_design.html#ts-apisurface-serviceclient

See the Client API example.
>  // client constructors have overloads for handling different
>  // authentication schemes.
>  constructor(connectionString: string, options?: ServiceClientOptions);
>  constructor(host: string, credential: TokenCredential, options?: ServiceClientOptions);

Reply:
The reason the example you posted is considered bad is because the type signatures for both constructors are exactly the same (string, options). In our case though we have 2 unique signatures (string, options) and (string, TokenCredential, options)

Reply:
Unable to reply to other comment for some reason :? 
Yes, EventHubs impl seems to have the different concerns addressed. The type of error being thrown is useful to note as well (TypeError as opposed to simple Error)
I just have the guideline aspect clarification pending - ok to be consistent with EventHubs if that's what we want to do
Thoughts? @ramya-rao-a @bterlson 

Personally, I think the example called out on the guidelines page resonates with my initial thoughts on switching the parameters - also because credential based constructor sounded more right too, than host based. (Going by the first parameter to pass in).

Reply:
Ok, the other observation I had was that options parameter being optional - seemed out of place to rely on type of to decide which constructor to use at first. So without that it can come down to two parameter constructor :?

Like, JS in general isn't strict on parameter numbers to be passed in.
So I think the one usecase that I had brought up earlier and seems to not be handled by EventHubs is if user invokes the host and credential based constructor as
(hostName, undefined, options) - then it looks like the error user would get would be about having to pass in a correctly formatted connection string.

Reply:
To summarize the thread here - the user experience for when credential is passed in as `undefined` is the one that seems unaddressed.

Let's take a call on if we want to address that as a bug and fix it here or not - and if we want to still go ahead with mirroring EventHubs constructors for most part. 

cc @ramya-rao-a @chradek 
@bterlson @AlexGhiondea

Reply:
The situation where a user passed undefined for the credential but still passes the options after it seems very unlikely. I think you could make the argument that if they're explicitly trying to not pass in a TokenCredential, then they would want a connection string (so the error seems relevant). But, I don't feel too strongly since they would be going against the types and docs in this case (I'm glad we give some error).

With regards to the parameter order for the constructors, my vote is for consistency. Since event-hubs is already GA, changing the order there would need a major version bump. I see one of our goals as providing a consistent experience across the packages we own (when possible) to lower the learning curve for users going from one package to another.

Reply:
Fair point, updating parameters ordering to keep up with consistency as if we decide to change, we can do it together

Agree that explicitly passing in undefined would map more to connection string based constructor.
But if say a user is instantiating the credentials programmatically, they could run into an `undefined` object. 
Event Hubs seems a bit different in that it accepts event hub name in addition to connection string in Event Hubs SDK. So it seems a little more relevant to Event Hubs as the err msg can be confusing to troubleshoot. (Also I wonder how long it might be before we bump version, maybe that would happen at same time as when Track 2 SB would go GA, but I digress :) )

And.. if we are of pinion that this a not so ideal state afterall, then we could have not so ideal string processing to guard against the missing/undefined token credentials case in Service Bus possibly. Let me know. For now haven't included those checks since we didn't want those. @ramya-rao-a Ok if we choose to handle it all together later as well or file a new issue? 
Because it seems like we might be regressing from Track 1 if we are no longer able to surface the error we used to be able to - ' "'credentials' is a required parameter and must be an instance of ... '